conans/client/graph/graph_builder.py

```diff
@@ -78,29 +79,34 @@ def extend_build_requires(self, graph, node, build_requires_refs, check_updates,
 
         conanfile = node.conanfile
         scope = conanfile.display_name
-        requires = [Requirement(ref) for ref in build_requires_refs]
-        for r in requires:
+
+        build_requires = []
+        for ref, context in build_requires_refs:
+            r = Requirement(ref)
             r.build_require = True
+            r.build_require_context = context
+            build_requires.append(r)
 
         if graph_lock:
             graph_lock.pre_lock_node(node)
-            graph_lock.lock_node(node, requires, build_requires=True)
+            # TODO: Add info about context?
+            graph_lock.lock_node(node, build_requires, build_requires=True)
 
-        self._resolve_ranges(graph, requires, scope, update, remotes)
+        self._resolve_ranges(graph, build_requires, scope, update, remotes)
 
-        for require in requires:
-            self._expand_require(require, node, graph, check_updates, update,
-                                 remotes, profile_host, new_reqs, new_options, graph_lock)
+        for br in build_requires:
+            context = br.build_require_context if node.context == CONTEXT_HOST else CONTEXT_BUILD
```

Comment:
I would keep the `node.context` here to show that there is no context switch once the build-requires is not consumed by a node in the _host_ context.

```suggestion
            context = br.build_require_context if node.context == CONTEXT_HOST else node.context
```